### DISCUSSION TITLE
List of experiments to try

### DISCUSSION THREAD

**Discussion by SuperUserNameMan** [dbuezas/lgt8fx#98] [Ideas]
- [ ] Execute machine code wrote into EEPROM (discussion [here](https://github.com/dbuezas/lgt8fx/discussions/98#discussioncomment-464322) )
- [ ] Drive a LED dot matrix directly using MiniEVB-LQFP32
- [ ] VGA output : http://www.gammon.com.au/forum/?id=11608 (discussion [here](https://github.com/dbuezas/lgt8fx/discussions/98#discussioncomment-463367) )
- [ ] "TinyBasicPlus" : https://github.com/BleuLlama/TinyBasicPlus (discussion [here](https://github.com/dbuezas/lgt8fx/discussions/98#discussioncomment-463392) )
- [ ] port "AVR Emulator of IC AY-8910" from: ATmega328p to LGT8Fx : https://www.avray.ru/ (discussion [here](https://github.com/dbuezas/lgt8fx/discussions/98#discussioncomment-473907) )
- [X] port Module sountrack player [PocketMod](https://github.com/rombankzero/pocketmod) to LGT8F328p : ( done [here] (https://github.com/dbuezas/lgt8fx/discussions/105) )
- [ ] Port "ArduinoShrink" library to LGT8F8x ? [ArduinoShrink](https://github.com/nerdralph/ArduinoShrink) : This library replaces Arduino AVR Core functions with smaller and faster versions, with the same or better functionality. The Arduino Blink example sketch built for the Uno takes 924 bytes of flash, but only needs 196 bytes with ArduinoShrink. ( discussion [here](https://github.com/dbuezas/lgt8fx/discussions/98#discussioncomment-570065) )
-------

## Execute machine code wrote into EEPROM

Because the EEPROM is emulated using part of the main Flash memory, maybe its possible to write machine code into it, and jump to its Flash address ?

When the EEPROM is configured to 1KB (default setting), it uses 2KB into the Flash because it swaps between an old 1KB page and a new 1KB page each time we write to the EEPROM (see explanations [here]( https://github.com/dbuezas/lgt8fx/pull/64 ) )

The first page is stored at 0x7800, and the second page at 0x7c00.

**Comment by SuperUserNameMan** [comment]
## VGA output

If a Atmega328p @ 16MHz can do that, there is no reason LGT8F328p @ 32MHz could not do slightly better :
http://www.gammon.com.au/forum/?id=11608

**Reply by jayzakk** [reply]
Reminds me of that very old Atari "computer" games. The hardware did not have a GPU, so the CPU had to push all data to screen directly. Damn, I am old....

**Comment by SuperUserNameMan** [comment]
## Tiny Basic Plus 

https://github.com/BleuLlama/TinyBasicPlus

Pretty impressive basic interpreter for Atmega328p

**Comment by jayzakk** [comment]
About using eeprom for code: would be great for program generated unrolled loops for speed.
But then, 1k is not that much... And maybe shouldn't write to eeprom that much :)

**Reply by SuperUserNameMan** [reply]
@jayzakk :

Yeah, also, since each AVR instruction is 16bit wide, the code would need to fit in 510 words.

Maybe it's possible to write useful routines in 510 words ... And if that's not enough, there is still the possibility to [change the size of the EEPROM](https://github.com/dbuezas/lgt8fx/wiki/EEPROM-of-the-LGT8F328p-:-the-important-things-you-need-to-know#h-2) at run-time.

We could have up to 8160 bytes (or 4080 words), but we would have to [skip the last 4 bytes of each 1K page](https://github.com/dbuezas/lgt8fx/wiki/EEPROM-of-the-LGT8F328p-:-the-important-things-you-need-to-know#4-for-each-1kb-of-emulated-eeprom-only-1020-are-available-) ....

**Comment by SuperUserNameMan** [comment]
## AVR Emulator of IC AY-8910

The AY-8910 was a very popular sound-chip used in many 8bits computers and consoles in the 80's.

This guy at https://www.avray.ru/ managed to emulate it on various ATmega. 

This video show comparison of real AY versus AVR-AY :
https://www.youtube.com/watch?v=Ii5gcyprhAg

**Comment by youxiaojie** [comment]
good idea!

**Comment by SuperUserNameMan** [comment]
## Porting ArduinoShrink

 [ArduinoShrink](https://github.com/nerdralph/ArduinoShrink) : This library replaces Arduino AVR Core functions with smaller and faster versions, with the same or better functionality. 

The Arduino Blink example sketch built for the Uno takes 924 bytes of flash, but only needs 196 bytes with ArduinoShrink.

Blog introduction : https://nerdralph.blogspot.com/2021/04/honey-i-shrunk-arduino-core.html
Github : https://github.com/nerdralph/ArduinoShrink
